rc/extra/racer.kak

```diff
@@ -93,3 +93,54 @@ define-command racer-disable-autocomplete -docstring "Disable racer completion"
     remove-hooks window racer-autocomplete
     unalias window complete racer-complete
 }
+
+define-command racer-go-definition -docstring "Jump to where the rust identifier below the cursor is defined" %{
+    %sh{
+        dir=$(mktemp -d "${TMPDIR:-/tmp}"/kak-racer.XXXXXXXX)
+        printf %s\\n "set-option buffer racer_tmp_dir ${dir}"
+        printf %s\\n "evaluate-commands -no-hooks %{ write ${dir}/buf }"
+    }
+    %sh{
+        dir=${kak_opt_racer_tmp_dir}
+        (
+            cursor="${kak_cursor_line} $((${kak_cursor_column} - 1))"
+            racer_data=$(racer --interface tab-text  find-definition ${cursor} ${kak_buffile} ${dir}/buf | head -n 1)
+            racer_match=$(printf %s\\n "$racer_data" | cut -f1 )
+            if [[ "$racer_match" == "MATCH" ]]; then
+              racer_line=$(printf %s\\n "$racer_data" | cut -f3 )
+              racer_column=$(printf %s\\n "$racer_data" | cut -f4 )
+              racer_file=$(printf %s\\n "$racer_data" | cut -f5 )
+              printf %s\\n "edit -existing '$racer_file' $racer_line $racer_column"
+              printf %s\\n "set-option buffer readonly true"
+            else
+              printf %s\\n "echo -debug 'racer could not find a definition'"
+            fi
+        )
+    }
+}
+
+define-command racer-show-doc -docstring "Show the documentation about the rust identifier below the cursor" %{
+    %sh{
+        dir=$(mktemp -d "${TMPDIR:-/tmp}"/kak-racer.XXXXXXXX)
+        printf %s\\n "set-option buffer racer_tmp_dir ${dir}"
+        printf %s\\n "evaluate-commands -no-hooks %{ write ${dir}/buf }"
+    }
+    %sh{
+        dir=${kak_opt_racer_tmp_dir}
+        (
+            printf -v NEWLINE '\n'
+            cursor="${kak_cursor_line} ${kak_cursor_column}"
+            racer_data=$(racer --interface tab-text  complete-with-snippet  ${cursor} ${kak_buffile} ${dir}/buf | head -n 2 | tail -n 1)
+            racer_match=$(printf %s\\n "$racer_data" | cut -f1)
+            printf %s\\n "info %@.$racer_match.@"
+            if [[ "$racer_match" == "MATCH" ]]; then
+              racer_doc=$(printf %s\\n "$racer_data" | cut -f9 | xargs )
```

Comment:
`xargs` isn't guaranteed to be installed on all systems, you'll have to remove it. I think this subshell can be replaced with an `awk` one-liner, but I'm not sure the behavior of `xargs` without any command is the same everywhere, so I don't know what the expected result is here.

Reply:
I'm guessing this whole `racer_doc` manipulation is supposed to do something like "remove all real newlines, unescape escaped newlines", yes? Then something like this should work:

    racer_doc=$(printf '%s\n' "$racer_data" | sed -E -e '
        # Grab the 9th tab-delimited field of every line
        s/([^\t]*\t){8}([^\t]*)/\2/
        # Copy the first line to the hold buffer
        1 h
        # Append the second and subsequent lines to the hold buffer
        2,$ H
        # When we get to the last line...
        $ {
            # Move the whole blob back to the pattern buffer
            x
            # Delete all the real newlines in the pattern buffer
            s/\n/ /g
            # Unescape all the escaped newlines
            s/\\n/\n/g
            # Escape any @ signs so we can use @ as a delimiter for Kakoune
            s/@/\\@/g
        }
    ')

Reply:
Unfortunately, `-E` isn't defined by POSIX.

Reply:
Dammit! I checked `sed(1)`, which says "the -E option has been supported for years by GNU sed, and is now included in POSIX.", but I failed to check `sed(1posix)` or [the online version](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/sed.html).

StackOverflow [points](https://unix.stackexchange.com/questions/310446/are-there-standard-versions-of-sed-on-which-e-is-unsupported) to [an issue in the POSIX issue-tracker](http://austingroupbugs.net/view.php?id=528) that adds `-E`, but although the issue was resolved at the end of 2012, apparently the change didn't make it into the 2016 revision.